pallets/parallel-xcm/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn xcm_fees)]
+    pub type XcmFees<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::storage]
+    #[pallet::getter(fn total_reserves)]
+    pub type TotalReserves<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Xcm message send failure
+        SendXcmError,
+    }
+}
+
+pub trait ParallelXCM<Balance, AssetId, AccountId> {
+    fn update_xcm_fees(fees: Balance);
+
+    fn update_total_reserves(reserves: Balance) -> DispatchResult;
+
+    fn ump_transact(
+        call: DoubleEncoded<()>,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+    ) -> Result<Xcm<()>, DispatchError>;
+
+    fn do_withdraw(
+        para_id: ParaId,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        para_account_id: AccountId,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+    ) -> Result<(), DispatchError>;
+
+    fn do_contribute(
+        para_id: ParaId,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetId,
+        account_id: AccountId,
+        amount: Balance,
+        xcm_fees_payer: AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+        who: &AccountId,
+    ) -> Result<(), DispatchError>;
+}
+
+impl<T: Config> ParallelXCM<BalanceOf<T>, AssetIdOf<T>, T::AccountId> for Pallet<T> {
+    fn update_xcm_fees(fees: BalanceOf<T>) {
+        XcmFees::<T>::mutate(|v| *v = fees);
+    }
+
+    fn update_total_reserves(reserves: BalanceOf<T>) -> DispatchResult {
+        TotalReserves::<T>::try_mutate(|b| -> DispatchResult {
+            *b = b.checked_add(reserves).ok_or(ArithmeticError::Overflow)?;
+            Ok(())
+        })
+    }
+
+    fn ump_transact(
+        call: DoubleEncoded<()>,
+        weight: Weight,
+        beneficiary: MultiLocation,
+        relay_currency: AssetIdOf<T>,
+        account_id: T::AccountId,
+        xcm_fees_payer: T::AccountId,
+        xcm_fees_payment_strategy: XcmFeesPaymentStrategy,
+    ) -> Result<Xcm<()>, DispatchError> {
+        let fees = Self::xcm_fees();
+        let asset: MultiAsset = (MultiLocation::here(), fees).into();
+
+        match xcm_fees_payment_strategy {
+            XcmFeesPaymentStrategy::Reserves => {
+                T::Assets::burn_from(relay_currency, &account_id, fees)?;
+
+                TotalReserves::<T>::try_mutate(|b| -> DispatchResult {
+                    *b = b.checked_sub(fees).ok_or(ArithmeticError::Underflow)?;
+                    Ok(())
+                })?;
```

Comment:
3/3 now decrease `TotalReserves` here, and do asset transferring inside this pallet, 
if it is a bug?

Reply:
Maybe here `account_id` should always be PalletId account, but I suggest that we can move this code inside into `T::XCM::update_total_reserves`
https://github.com/parallel-finance/parallel/blob/2d5382a471e2a0ce0b04579af662f37830d23fb9/pallets/crowdloans/src/lib.rs#L500-L506

Reply:
@MrPai Regarding these comments, Please create separate issue. As a part of this PR, I just moved XCM related functions from Crowdloan to XCM Helper. 

For ex: I moved entire this methid https://github.com/parallel-finance/parallel/blob/master/pallets/crowdloans/src/lib.rs#L657 

cc @ETeissonniere

Reply:
yes, please close this issue #1002 once finished

Reply:
@MrPai I am marking this comment as resolved. #1002 would be done separately. and this is not related to this PR. 

cc @ETeissonniere